text/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,475 @@
+how well the current design satisfies these constraints.
+
+* **Changing from `panic=unwind` to `panic=abort` cannot cause undefined
+  behavior:** We wish to ensure that changing from `panic=unwind` to
+  `panic=abort` never creates undefined behavior (relate to `panic=unwind`),
+  even if one is relying on a library that triggers a panic or a foreign
+  exception.
+* **Optimization with `panic=abort`:** when using `panic=abort`, we
+  wish to enable as many code-size optimizations as possible. This
+  means that we shouldn't have to generate unwinding tables or other
+  such constructs, at least in most cases.
+* **Preserve the ability to change how Rust panics are propagated when
+  using the Rust ABI:** Currently, Rust panics are propagated using
+  the native unwinding mechanism, but we would like to keep the
+  freedom to change that.
+* **Enable Rust panics to traverse through foreign frames:** Several
+  projects would like the ability to have Rust panics propagate
+  through foreign frames.  Those frames may or may not register
+  destructors of their own with the native unwinding mechanism.
+* **Enable foreign exceptions to propagate through Rust frames:**
+  Similarly, we would like to make it possible for C++ code (or other
+  languages) to raise exceptions that will propagate through Rust
+  frames "as if" they were Rust panics (i.e., running destrutors or,
+  in the case of `unwind=abort`, aborting the program).
+* **Enable error handling with `longjmp`:**
+  As mentioned above, some existing Rust libraries rely on the ability to
+  `longjmp` across Rust frames to interoperate with Ruby, Lua, and other C
+  APIs. The behavior of `longjmp` traversing Rust frames is not specified or
+  guaranteed to be safe; in the current implementation of `rustc`,
+  however, it [is safe][longjmp-pr]. On Windows, `longjmp` is implemented as a
+  form of unwinding called ["forced unwinding"][forced-unwinding], so any
+  specification of the behavior of forced unwinding across FFI boundaries
+  should be forward-compatible with a [future RFC][unresolved-questions] that
+  will provide a well-defined way to interoperate with longjmp-based APIs.
+* **Do not change the ABI of functions in the `libc` crate:** Some `libc`
+  functions may invoke `pthread_exit`, which uses [a form of
+  unwinding][forced-unwinding] in the GNU libc implementation. Such functions
+  must be safe to use with the existing `"C"` ABI, because changing the types
+  of these functions would be a breaking change. 
+
+[inside-rust-requirements]: https://blog.rust-lang.org/inside-rust/2020/02/27/ffi-unwind-design-meeting.html#requirements-for-any-cross-language-unwinding-specification
+[longjmp-pr]: https://github.com/rust-lang/rust/pull/48572
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Functions declared with `extern "C"` are generally not permitted to unwind
+(with some very narrow exceptions, described [below][forced-unwinding]).
```

Comment:
"Not permitted" here means "doing so causes UB", right? That might be worth stating explicitly.

Reply:
Yes, unless it's a `panic`, in which case the process aborts. I agree this could be clearer, but upon re-reading it, I'm not sure this paragraph is even necessary.